typing/typecore.ml

```diff
@@ -1803,108 +1805,117 @@ let rec is_nonexpansive exp =
         ) p;
         !res
       in
-      is_nonexpansive e &&
-      List.for_all
+      is_expansive e ||
+      List.exists
         (fun {c_lhs; c_guard; c_rhs} ->
-           is_nonexpansive_opt c_guard && is_nonexpansive c_rhs
-           && not (contains_exception_pat c_lhs)
+           is_expansive_opt c_guard ||
+           is_expansive c_rhs ||
+           contains_exception_pat c_lhs
         ) cases
-  | Texp_tuple el ->
-      List.for_all is_nonexpansive el
+  | Texp_tuple el
   | Texp_construct( _, _, el) ->
-      List.for_all is_nonexpansive el
-  | Texp_variant(_, arg) -> is_nonexpansive_opt arg
+      List.exists is_expansive el
+  | Texp_variant(_, arg) -> is_expansive_opt arg
   | Texp_record { fields; extended_expression } ->
-      Array.for_all
+      Array.exists
         (fun (lbl, definition) ->
            match definition with
            | Overridden (_, exp) ->
-               lbl.lbl_mut = Immutable && is_nonexpansive exp
+               lbl.lbl_mut = Mutable || is_expansive exp
            | Kept _ -> true)
```

Comment:
` Kept _ -> false` ? Since keeping all labels should be non-expansive.

Reply:
Indeed! I missed that one. Thanks for the careful read :)

Reply:
Fixed.